### PR TITLE
🌱 Update golangci-lint to latest

### PR DESCRIPTION
**Pull request by furkatgofurov7**
**What this PR does / why we need it**:
Golangci-lint is failing intermittently in a few PRs':

- https://prow.apps.ci.metal3.io/view/gcs/metal3-prow-artifacts/pr-logs/pull/metal3-io_cluster-api-provider-metal3/175/golint/1366761909239418884
- https://prow.apps.ci.metal3.io/view/gcs/metal3-prow-artifacts/pr-logs/pull/metal3-io_cluster-api-provider-metal3/175/golint/1366761909247807489
- https://prow.apps.ci.metal3.io/view/gcs/metal3-prow-artifacts/pr-logs/pull/metal3-io_cluster-api-provider-metal3/157/golint/1366744921666490372
- https://prow.apps.ci.metal3.io/view/gcs/metal3-prow-artifacts/pr-logs/pull/metal3-io_cluster-api-provider-metal3/171/golint/1366751339048079365

This patch updates it to the latest  as the speed of the latest version seems to be increased a lot per
[golangci/golangci-lint#685 (comment).](https://github.com/golangci/golangci-lint/issues/685#issuecomment-541886167)

~~Since we moved to go1.16 in #161 we were hitting a go bug (see golint prow job output below, when a workaround is not applied ) for how go build works with modules - [golang/go#44129](https://github.com/golang/go/issues/44129)~~

```
` + make lint
cd hack/tools; go build -tags=tools -o bin/golangci-lint github.com/golangci/golangci-lint/cmd/golangci-lint
go: github.com/golangci/golangci-lint@v1.37.1: missing go.sum entry; to add it:
	go mod download github.com/golangci/golangci-lint
make: *** [Makefile:124: hack/tools/bin/golangci-lint] Error 1 `
```

~~This commit adds the **-mod=mod** flag for **go build** command in the Makefile ([workaround](https://github.com/golang/go/issues/44129#issuecomment-789250634) mentioned in the issue).~~

### CODE REVIEW

**Comment by fmuyassarov** [comment]
I think you need to run `go mod tidy` after updating go.mod.

**Review by namnx228 — COMMENTED** [review]
LGTM

**Comment by furkatgofurov7** [comment]
> I think you need to run `go mod tidy` after updating go.mod.

Special thanks @fmuyassarov, it fixed.

**Comment by furkatgofurov7** [comment]
/hold
Similar PR in IPAM did not fail and is fine without **-mod=mod** workaround, investigating.

**Comment by furkatgofurov7** [comment]
/hold cancel
We can go without a workaround in this patch as well, something was messing up in the previous prow runs that made linting to fail.

File: hack/tools/go.mod [hunk 1]
```diff
@@ -3,26 +3,10 @@ module sigs.k8s.io/cluster-api/hack/tools
...
+	github.com/golangci/golangci-lint v1.37.1
+	github.com/jteeuwen/go-bindata v3.0.7+incompatible
 	k8s.io/code-generator v0.18.2
 	sigs.k8s.io/controller-tools v0.3.0
-	sigs.k8s.io/structured-merge-diff v0.0.0-20190525122527-15d366b2352e // indirect
+	sigs.k8s.io/kustomize/kustomize/v3 v3.10.0
 	sigs.k8s.io/testing_frameworks v0.1.1
 )
-
-replace (
```
Comment (furkatgofurov7) [inline hack/tools/go.mod:16]: removing it: [golangci/golangci-lint#595](https://github.com/golangci/golangci-lint/issues/595) is fixed.

**Comment by kashifest** [comment]
/approve
I can see in the comment history that you needed to run golint test again in this PR as well, so it is solving anything or is it just uplifting the version.

File: hack/tools/go.mod [hunk 2]
```diff
@@ -3,26 +3,10 @@ module sigs.k8s.io/cluster-api/hack/tools
...
-	github.com/drone/envsubst v1.0.3-0.20200804185402-58bc65f69603 // indirect
-	github.com/golang/mock v1.4.3
-	github.com/golangci/golangci-lint v1.17.1
-	gonum.org/v1/netlib v0.0.0-20190331212654-76723241ea4e // indirect
+	github.com/golangci/golangci-lint v1.37.1
+	github.com/jteeuwen/go-bindata v3.0.7+incompatible
 	k8s.io/code-generator v0.18.2
 	sigs.k8s.io/controller-tools v0.3.0
-	sigs.k8s.io/structured-merge-diff v0.0.0-20190525122527-15d366b2352e // indirect
+	sigs.k8s.io/kustomize/kustomize/v3 v3.10.0
```
Comment (fmuyassarov) [inline hack/tools/go.mod:10]: Would you mind to share why do we need it here?

**Comment by fmuyassarov** [comment]
@furkatgofurov7 would you mind please to rebase the PR so that we could move on ?

File: hack/tools/go.mod (hunk 2 above)
Comment (fmuyassarov) [reply hack/tools/go.mod:10]: or was the plan to use the kustomize directly from go.mod ?

Comment (furkatgofurov7) [reply hack/tools/go.mod:10]: It seems to autogenerate it while running _go mod tidy_, will check that.

**Comment by fmuyassarov** [comment]
@kashifest @furkatgofurov7 
I'm not sure how we ended up using gofmt instead of golint. But using golint image from quay instead of golang and replacing `hack/gofmt` with `hack/golint.sh` seems to fix the issue.

**Comment by fmuyassarov** [comment]
> @kashifest @furkatgofurov7
> I'm not sure how we ended up using gofmt instead of golint. But using golint image from quay instead of golang and replacing `hack/gofmt` with `hack/golint.sh` seems to fix the issue.

Quay is building the golint image from https://github.com/metal3-io/baremetal-operator/blob/e74c3d8461e2e5f55d3b0839988d04c254490c1e/hack/Dockerfile.golint

**Comment by furkatgofurov7** [comment]
/test-centos-integration

> @kashifest @furkatgofurov7
> I'm not sure how we ended up using gofmt instead of golint. But using golint image from quay instead of golang and replacing `hack/gofmt` with `hack/golint.sh` seems to fix the issue.

True, I have not noticed it before, and this should be a bug, IMO? But if just changing to correct image helps, we do not need to bump golangci-lint version at all :/

File: hack/tools/go.mod [hunk 3]
```diff
@@ -12,18 +12,4 @@ require (
 	sigs.k8s.io/controller-tools v0.5.0
 	sigs.k8s.io/structured-merge-diff v0.0.0-20190525122527-15d366b2352e // indirect
 	sigs.k8s.io/testing_frameworks v0.1.1
-)
```
Comment (fmuyassarov) [inline hack/tools/go.mod:15]: @furkatgofurov7 Probably this was removed accidentally?

File: hack/tools/go.mod [hunk 4]
```diff
@@ -12,18 +12,4 @@ require (
...
-	github.com/golangci/errcheck v0.0.0-20181003203344-ef45e06d44b6 => github.com/golangci/errcheck v0.0.0-20181223084120-ef45e06d44b6
-	github.com/golangci/go-tools v0.0.0-20180109140146-af6baa5dc196 => github.com/golangci/go-tools v0.0.0-20190318060251-af6baa5dc196
-	github.com/golangci/gofmt v0.0.0-20181105071733-0b8337e80d98 => github.com/golangci/gofmt v0.0.0-20181222123516-0b8337e80d98
-	github.com/golangci/gosec v0.0.0-20180901114220-66fb7fc33547 => github.com/golangci/gosec v0.0.0-20190211064107-66fb7fc33547
-	github.com/golangci/ineffassign v0.0.0-20180808204949-42439a7714cc => github.com/golangci/ineffassign v0.0.0-20190609212857-42439a7714cc
-	github.com/golangci/lint-1 v0.0.0-20180610141402-ee948d087217 => github.com/golangci/lint-1 v0.0.0-20190420132249-ee948d087217
-	github.com/timakin/bodyclose => github.com/golangci/bodyclose v0.0.0-20190714144026-65da19158fa2
-	mvdan.cc/unparam v0.0.0-20190124213536-fbb59629db34 => mvdan.cc/unparam v0.0.0-20190209190245-fbb59629db34
-)
+)
```
Comment (furkatgofurov7) [inline hack/tools/go.mod:15]: @fmuyassarov this replaces it.

Comment (fmuyassarov) [reply hack/tools/go.mod:15]: ah right, missed that part

**Comment by furkatgofurov7** [comment]
Tackled in #206 

/close

**Pull request closed without merging**